backend/stellar_refund_validator.go

```diff
@@ -0,0 +1,111 @@
+	Type:   "refund_already_executed",
+	Title:  "Refund Already Executed",
+	Status: http.StatusBadRequest,
+	Detail: "The refund has already been executed.",
+}
+
+// StellarRefundValidator checks if it is possible to
+// refund a deposit to depositor's Stellar account.
+type StellarRefundValidator struct {
+	Session                db.SessionInterface
+	WithdrawalWindow       time.Duration
+	Observer               ethereum.Observer
+	EthereumFinalityBuffer uint64
+}
+
+// StellarRefundDetails includes metadata about the
+// validation result.
+type StellarRefundDetails struct {
+	// LedgerSequence is the sequence number of the Stellar ledger
+	// for which the validation result is accurate.
+	LedgerSequence uint32
+}
+
+func (s StellarRefundValidator) CanRefund(ctx context.Context, deposit store.StellarDeposit) (StellarRefundDetails, error) {
+	dbStore := store.DB{Session: s.Session.Clone()}
+	err := dbStore.Session.BeginTx(&sql.TxOptions{
+		Isolation: sql.LevelRepeatableRead,
+		ReadOnly:  true,
+	})
+	if err != nil {
+		return StellarRefundDetails{}, errors.Wrap(err, "error starting repeatable read transaction")
+	}
+	defer func() {
+		// explicitly ignore return value to make the linter happy
+		_ = dbStore.Session.Rollback()
+	}()
+
+	lastLedgerSequence, err := dbStore.GetLastLedgerSequence(context.Background())
+	if err != nil {
+		return StellarRefundDetails{}, errors.Wrap(err, "error getting last ledger sequence")
+	}
+
+	// Checks on Stellar side:
+	// - Ensure that currently ingested ledger is past the withdrawal window
+	// - Ensure that refund tx has not been already executed
+	lastLedgerCloseTime, err := dbStore.GetLastLedgerCloseTime(ctx)
+	if err != nil {
+		return StellarRefundDetails{}, errors.Wrap(err, "error getting last ledger close time")
+	}
+	withdrawalDeadline := time.Unix(deposit.LedgerTime, 0).Add(s.WithdrawalWindow)
+	if !lastLedgerCloseTime.After(withdrawalDeadline) {
+		return StellarRefundDetails{}, WithdrawalWindowStillActive
+	}
+
+	// Check if refund tx was seen without signature request
+	exists, err := dbStore.HistoryStellarTransactionExists(ctx, deposit.ID)
+	if err != nil {
+		return StellarRefundDetails{}, errors.Wrap(err, "error getting history stellar transaction by memo hash")
+	}
+	if exists {
+		return StellarRefundDetails{}, RefundAlreadyExecuted
+	}
+
+	// rollback to release used DB connection because further checks
+	// do not involve DB
+	_ = dbStore.Session.Rollback()
+
+	// Checks on Ethereum side:
+	// - Ensure that there was no withdrawal to Ethereum account
+	// - The response from the client is after the withdrawal deadline
+	depositID := common.HexToHash(deposit.ID)
+	requestStatus, err := s.Observer.GetRequestStatus(ctx, depositID)
+	if err != nil {
+		return StellarRefundDetails{}, errors.Wrap(err, "error getting request status from ethereum observer")
+	}
+
+	block, err := s.Observer.GetBlockByNumber(ctx, requestStatus.BlockNumber-s.EthereumFinalityBuffer)
```

Comment:
we can add an extra check that `requestStatus.BlockNumber` is greater than `s.EthereumFinalityBuffer` to prevent underflow